### ISSUE TITLE
Orientation/symmetry issues in Box2BoxTransformRotated.get_deltas

### ISSUE THREAD

**Issue by alekseynp** [facebookresearch/detectron2#1172]
I believe I have identified two issues with Box2BoxTransformRotated that are related to notions of orientation and symmetry.

During training, boxes are matched based on their pairwise rotated iou, however the target regression deltas can end up large with the current logic.

I illustrate my point with the following test cases:

```python
import torch

from detectron2.structures import RotatedBoxes, pairwise_iou_rotated
from detectron2.modeling.box_regression import Box2BoxTransformRotated

box2box_transform = Box2BoxTransformRotated(weights=(1, 1, 1, 1, 1))

# Failure 1

# These are the same boxes, just flipped
box1 = RotatedBoxes(torch.tensor([[40.0, 40.0, 40.0, 5.0, 0]]))
box2 = RotatedBoxes(torch.tensor([[40.0, 40.0, 40.0, 5.0, 180]]))

# tensor([[1.]])
pairwise_iou_rotated(box1, box2)

# tensor([[ 0.0000,  0.0000,  0.0000,  0.0000, -3.1416]])
box2box_transform.get_deltas(box1.tensor, box2.tensor)

# Failure 2

# These boxes lie on top of each other, just transposed
box1 = RotatedBoxes(torch.tensor([[40.0, 40.0, 40.0, 5.0, 0]]))
box2 = RotatedBoxes(torch.tensor([[40.0, 40.0, 5.0, 40.0, 90]]))

# tensor([[1.]])
pairwise_iou_rotated(box1, box2)

# tensor([[ 0.0000,  0.0000, -2.0794,  2.0794,  1.5708]])
box2box_transform.get_deltas(box1.tensor, box2.tensor)
```

In both of the above failure cases, based on iou box1 would be matched to box2 as a perfect match, but the target regression deltas are extremely large. I have seen this negatively impact training.

**Failure 1**

If we do not consider a bounding box to have an orientation, then a rotation of -110 degrees is the same as a rotation of +70 degrees. If you do consider it to have an orientation, then the +70 degree rotated box is "upside down" and incorrect.

The current Box2BoxTransformRotated.get_deltas logic maps angles to [180, 180). I assert this interprets boxes to have a orientation. The orientation invariant version of the logic would be the same, except angles would be mapped to [90, 90).

**Failure 2**

If we do not consider boxes to have an orientation, then the delta between these boxes is zero, but if we do not, then we must rotate the box 90 degrees and dramatically change its width and height.

I would propose that Box2BoxTransformRotated.get_deltas should consider two possibilities:
1. That the source width, and height should be matched to the target width and height
2. That the source width should be mapped to the target height, and vice versa with a rotation of 90 degrees.

Whichever delta is "smaller" should be considered the target. This is ambiguous because it depends on how you are inclined to trade off changes in width, height, and angle. I would propose to use the weights and the smooth L1 loss to inform this.

## Bigger picture

I think rotated bounding boxes are an incredibly valuable, but I see the complexity that they bring. Here this analysis is leading to greater complexity. I believe that there are ultimately two types of rotated bounding box. One with orientation and one without where that is ultimately defined by the ground truth labels and the use case.

Consider labelling people:
- Suppose that I always draw a rotated bounding box such that height is along the long axis of the human body, and the "up" end of the box is always the head. Suppose I also want oriented bounding box detections that align with this definition. Then I would like the entire detectron2 pipeline to respect that.
- Suppose instead I draw my bounding boxes without these considerations and I do not care about oriented detections. Then I want the entire pipeline to ignore orientation.

Some thorough analysis of this already exists in the project here:
https://github.com/facebookresearch/detectron2/blob/b8f4eebeb06827d49e31738c0997f919dac26aba/detectron2/layers/nms.py#L39

I would propose that as the vast majority of the logic is currently orientation-free and that Box2BoxTransformRotated.get_deltas should be changed accordingly.

I have a PR in development that I would be happy to raise after discussion of this issue.

In the long term, perhaps rotated bounding boxes should be split into the two types if the need arises.

**Comment by ppwwyyxx** [comment]
Our rotated models are designed to expect rotated boxes to have orientation. The main use is to detect texts, btw.
Under such definition, it does have some issues you mentioned like ill-defined IOU. We also discussed such issues in the code as you saw. It may cause some potential negative impact in situations like matching proposals to ground truth, but that's just what we have now and it works OK.
Contributions to improve these issues (or to add another type of rotated boxes) sounds useful. Unfortunately we don't currently have public benchmarks for these models so it would be hard for us to evaluate such contributions.

**Comment by alekseynp** [comment]
In [Arbitrary-Oriented Scene Text Detection via Rotation Proposals](https://arxiv.org/abs/1703.01086) they suppress high IoU positive matches in training if the angle difference is greater than pi/12. That is not simply a magic hyper-parameter, but is deliberately half their anchor spacing. In this way they have a somewhat crude. but effective oriented-IoU score. Or instead one could view it as a match scorer. In detectron2 this would be equivalent to setting elements of the match_quality_matrix to zero based on a angle difference threshold.

In that same paper they measure performance on benchmarks: MSRA-TD500, ICDAR2015, and ICDAR2013.

I would propose:
- Adding one or more public text detection data sets to function as a benchmark for oriented rotated boxes
- Generally implement the Arbitrary-Oriented Scene Text Detection via Rotation Proposals paper. There should be relatively little to do
- Add angle difference thresholding for match scores and compare
- Find a unoriented rotated box benchmark for the project
- Implement a Box2BoxTransformRotatedUnoriented that computes deltas without orientation
- Measure performance against the unoriented benchmark

**Comment by cbasavaraj** [comment]
I tried thresholding on delta of pi / 12, delta being the lower of a) absolute difference between GT and anchor, and b) absolute difference between GT and anchor rotated by 180. Didn't get better results though. One problem is the discontinuity (for the box regressor) I introduce by using anchor rotated by 180. But the alternative would be to use 13 angles instead of 7. 

Overall the results were ok, but the angle prediction kinda unstable, i.e., in the output video, rotating around from frame to frame. I think part of the problem was that our annotations were not oriented. 

Will come back to this later. Plan to jointly train mask segmentation + rotated boxes. In another thread, someone said this gave better results.

**Comment by alekseynp** [comment]
I think if you're going to do oriented matching, you must also have oriented anchors and oriented annotations. So, if by going from 7 angles to 13 angles you mean having anchors that cover 360 degrees rather than 180, yes I think this is necessary.

Are you aware of any instance segmentation data sets with (oriented?) rotated boxes?

Finally if everything is oriented, probably evaluation should be as well. A pi/12 rule or similar should be applied when matching in the evaluator.

**Comment by cbasavaraj** [comment]
Hi @alekseynp , no I'm not aware of any such datasets. 

Minor correction to what I said, it'll be 12 angles, not 13, for covering the 360 degrees like you say. For training, will the pi / 12 criterion really help? What I saw on Tensorboard is that both RPN and Fast RCNN had far more positive anchors when I didn't use the pi / 12 criterion. What do we care if the network can successfully learn to regress anchors that are far away (in orientation). 

Agree about evaluation, gotta figure out a way to combine orientation with IoU.

**Comment by alekseynp** [comment]
We broadly attempt to structure our models to make them smooth, stable, learn-able, etc. However I agree that if it can be empirically demonstrated to be irrelevant, then it doesn't matter.

The symptom I've seen is unstable regressed deltas. During training the model is sometimes told to change angles by small amounts and sometimes large amounts. During testing, the model sometimes has very large errors when it mistakenly behaves in the large-delta situation.

A potential data set is used here: https://github.com/mrlooi/rotated_maskrcnn
"PPC (Pens, Pencils, Chopsticks)"
https://drive.google.com/open?id=1B4jV49KGVJtiZJVxcG11kNzGQKV02NHz

Maybe this: https://captain-whu.github.io/DOTA/index.html

MAP Evaluation involves matching predictions to ground truth based on IoU. Easily an angle threshold like pi/12 could be introduced to the matching criteria to create an hard-orientation-threshold MAP.

Alternatively... Take the relative angle between two rotated boxes. Take the cosine. That is now on `[-1, 1]`. Linearly map that to `[0, 1]` with `0.5(x + 1)` and multiply it by IoU. Now we have softer oriented IoU.

**Comment by pfollmann** [comment]
@alekseynp you might have a look at our paper "Oriented Boxes for Accurate Instance Segmentation" (https://arxiv.org/abs/1911.07732). 
We modified the anchor-assignment in some ways to address the issues you have mentioned above (and some more) in order to stabilize the training. In particular, the use of ArIoU is also makes the training more efficient than calculating the exact IoU. (no cosine needed)
So far, I didn't have a closer look at the rotated box implementations here, but it's interesting to see that more and more people seem to start using oriented boxes.

**Comment by cbasavaraj** [comment]
@pfollmann Your paper looks great! Do you plan to integrate your code into Detectron2? Maybe on a fork? Thanks
